accepted/wasmtime-baseline-compilation.md

```diff
@@ -0,0 +1,260 @@
+otherwise (emitting with vregs first and editing later) requires actually
+buffering the MachInst structs in memory, which we do not wish to do.
+
+We don’t expect to make any changes to Cranelift itself beyond the layering
+refactor to borrow its `MachInst` and `MachBuffer` implementations. In
+particular we don’t expect to use the `MacroAssembler` wrapper in Cranelift, at
+least at first, because it will be built around constructing and emitting
+instructions to machine code right away, without buffering (as in Cranelift’s
+VCode). It’s possible in the future that we may find other ways to make
+`MacroAssembler` generic and leverage it in Cranelift too, but that is beyond
+the scope of this RFC.
+
+### Register Allocation
+
+We plan to implement register allocation in a linear[^2] fashion.
+
+[^2]: Known as [Linear Scan Register
+Allocation](http://web.cs.ucla.edu/~palsberg/course/cs132/linearscan.pdf)
+
+The baseline compiler will hold a reference to a register allocator abstraction,
+which will keep a list of registers, represented by Cranelit's `Reg`
+abstraction,  per ISA, along with their availability. It will also hold
+a reference to a value stack abstraction, to keep track of operands and results
+and their location as it performs compilation. These are the two key
+abstractions for register allocation:
+
+```rust
+pub struct Compiler {
+  //...
+  allocator: RegisterAllocator,
+  value_stack: ValueStack,
+  //...
+}
+```
+
+The value stack is expected to keep track of the location of its values.
+A particular value can be tagged as either a:
+
+* Local: representing a function local slot (index and type). The address of the
+  local will be resolved lazily to reduce register pressure. 
+* Register
+* Constant: representing an immediate value. 
+* Memory Offset: the location of the value at a given memory offset
+
+Registers will be requested to the register allocator every time an operation
+requires it. If no registers are available, the baseline compiler will move
+locals and registers to memory, changing their tag to a memory offset,
+performing what's known as spilling, effectively freeing up registers. Spilling
+will also be performed at control flow points. To reduce the number of spills,
+the baseline compiler will also perform limited constant rematerialization.
+
+Assuming that we have an immediate at the top of the stack, emitting an add
+instruction with an immediate operand would look something like this:
+
+```rust
+let mut masm = cranelift_asm::x64::MacroAssembler::new();
+// request a general purpose register;
+// spill if
+let imm = self.value_stack.pop(); 
+none available let rd = self.gpr();
+masm.add(rd, imm);
+```
+
+### Integration with Wasmtime
+
+We plan to integrate the baseline compiler incrementally into Wasmtime, as an
+in-tree crate, `winch`. It will be introduced as a runtime feature, off by
+default[^3]. Taking as a guideline [Wasmtime's tiers of
```

Comment:
I'm not convinced that making this a compile time feature will be overly complex, and even if it turns out that it is, I would strongly prefer that we start by attempting to do it that way, and only fall back to runtime-only if we determine it isn't practical.

Reply:
I see the appeal of compiled-out-by-default for minimizing impact, but I'll offer a few reasons from experience that might push in the other direction:

- Dealing with features in general is a speedbump for users. When developing the new x64 backend we had it cfg'd off behind a feature and it was a constant fight to get tools (LSP etc) to use the right flags, and meant that we couldn't have just one `cargo test` invocation to test the thing; we had to have a separate one with the feature enabled.
- Differential testing across the two configurations becomes impossible. For the x64 backend, I eventually had to refactor to use a runtime setting and allow both configs to co-exist so that I could build a differential fuzz target. In this case I guess the feature would be strictly additive (include another compiler backend) but I could imagine tests at the Wasmtime level might still run into issues.
- In general, features make for more confusing code; see e.g. @alexcrichton 's push for us to use very coarse-grained features in memfd, and remove the cfg-logic from the inner implementations; and also the difficulties arising from the feature-controlled uffd backend.

On the other side, I guess the main concern is compile time? Or are you worried about e.g. users prematurely opting into using it and that opt-in being easier if it's a programmatic configuration vs. a feature flag in the dependency?

Reply:
My concerns are (1) compile time and (2) factoring code such that it isn't entangled with the rest of Cranelift/Wasmtime.

Differential testing should definitely be possible. I'm _not_ saying that we should have a compile time config to either use Winch xor cranelift (like the `experimental_x64` cargo feature did). I'm saying we should have a compile time config which when enabled adds support for and exposes the Winch compilation strategy variant. So both Winch and Cranelift would be available for differential testing or what have you when the cargo feature was enabled.

I don't think this is an onerous imposition on development, and we can point to the ongoing component model work as something that is using this compile-time-cargo-feature-to-make-available-a-runtime-option approach successfully.

Reply:
Fair enough; I think much of my experience is with the "xor" variant of feature (so a bunch of `#[cfg(not(feature = ...))]`) and that may have given me a darker view of how this could go.

I do like the idea of ensuring proper layering/separation by having a build with the feature off, as well.

So I guess this is fine; as you say we can always abandon the approach and go runtime-only if needed later.

Reply:
I think I didn't do a good job expressing myself during the initial draft of the doc when I introduced the idea of/open questions around the compile time feature, which probably hinted the "xor" approach. I'm aligned with this though, my main goal is to allow the usage of both compilers, for ease of testing, but also I can definitely see cases in which an embedder might want to use Winch for fast startup and then compile with Cranelift in the background for subsequent requests; and according to the discussion here, such goal is still achievable with a compile time feature.

Reply:
I've updated the doc to specify that changes will be introduced behind a compile-time feature.